### PR TITLE
Feature/batch norm to semantic similarity

### PR DESCRIPTION
**Pull request by aCampello**
Description
---

Towards fine tuning linkage models (https://trello.com/c/9ZrYNkko) I added the flexibility to:

1. Turn dropout on and off for Semantic Similarity
2. Add batch norm

I also brought a method (`.initialise_models()`) previously private, to public, because users might want to initialise it for example for plotting.

These changes allow us to go from model 1:

Screenshot 2020-10-16 at 16 36 17

to model 2:

Screenshot 2020-10-16 at 16 02 32

Checklist
---

- [x] Added link to Github issue or Trello card
- [x] Added tests (Re-purposed a test)

### CODE REVIEW

File: tests/transformers/test_semantic_similarity.py [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
     SemanticEquivalenceClassifier, SemanticEquivalenceMetaClassifier
 
 
+# Random seeds for test reproducibility
+# Based on keras docs
+# https://keras.io/getting_started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development
+
+np.random.seed(42)
+python_random.seed(42)
+tf.random.set_seed(42)
```
Comment (nsorros) [inline tests/transformers/test_semantic_similarity.py:18]: nice

File: wellcomeml/ml/bert_semantic_equivalence.py [hunk 2]
```diff
@@ -240,19 +242,25 @@ class SemanticEquivalenceMetaClassifier(SemanticEquivalenceClassifier):
     and model initialisation. Fit, predict, score remain intact
     """
 
-    def __init__(self, n_numerical_features, dropout_rate=0.1, **kwargs):
+    def __init__(self, n_numerical_features, dropout=False,
+                 batch_norm=False, dropout_rate=0.1, **kwargs):
```
Comment (nsorros) [inline wellcomeml/ml/bert_semantic_equivalence.py:246]: maybe you want to default to dropout 0 to maintain backwards compatibility

**Review by nsorros — APPROVED** [review]
LGTM

**Comment by nsorros** [comment]
What is the rational behind adding batch norm?

**Comment by aCampello** [comment]
> What is the rational behind adding batch norm?

When you add new numerical features, those might be out of scale, which forces us to normalise the features before. Batchnorm helps with this.

File: wellcomeml/ml/bert_semantic_equivalence.py (hunk 2 above)
Comment (aCampello) [reply wellcomeml/ml/bert_semantic_equivalence.py:246]: That's what `dropout=False` does

**Comment by aCampello** [comment]
One the tests fails - investigating

**Comment by aCampello** [comment]
Hah - it seems that with batch norm 3 epochs is not enough for the loss to go down...

**Comment by aCampello** [comment]
[Fine-tune models for linkage](https://trello.com/c/9ZrYNkko/1121-fine-tune-models-for-linkage)

File: wellcomeml/ml/bert_semantic_equivalence.py (hunk 2 above)
Comment (nsorros) [reply wellcomeml/ml/bert_semantic_equivalence.py:246]: maybe you do not need two parameters then :P dropout 0 means no dropout is applied

Comment (aCampello) [reply wellcomeml/ml/bert_semantic_equivalence.py:246]: You are right. Maybe a merged too quick. In the next PR I will patch this change in.
